zdm/grid.py

```diff
@@ -365,18 +373,102 @@ def calc_pdv(self, beam_b=None, beam_o=None):
                 # the below is to ensure this works when w is a vector of length nz
                 w = np.array(w)
                 
-                self.b_fractions[:, :, i] += (
-                        self.beam_o[i]
+                # this array gives the relative probability of detecting an FRB at this point
+                # in the beam, with this particular width. We may not have space to store this
+                # as a 4D (w,b,z,DM) array, hence we store two 3D arrays
+                temp_wb = self.beam_o[i] \
                         * (self.array_cum_lf(
                             thresh, Emin, Emax, self.state.energy.gamma, self.use_log10
                         ).T * w.T).T
-                    )
+                
```

Comment:
have you tested whether this changed code gives the same results as before (if run the same way)?